File: client/sdk/golang/pkg/metrics/metrics.go
package metrics

import (
	"fmt"
	"io"
)

type RequestMetrics struct {
	Provider         string
	Model            string
	BaseURL          string
	Stream           bool
	ElapsedSec       float64
	TextChars        int
	ImageCount       int
	Status           string
	ExceptionMessage string
}

func Print(w io.Writer, m RequestMetrics) {
	elapsed := m.ElapsedSec
	if elapsed <= 0 {
		elapsed = 1e-9
	}
	tps := float64(m.TextChars) / elapsed

	fmt.Fprintln(w, "\n=== Request Metrics ===")
	fmt.Fprintf(w, "provider: %s\n", m.Provider)
	fmt.Fprintf(w, "model: %s\n", m.Model)
	fmt.Fprintf(w, "base_url: %s\n", m.BaseURL)
	fmt.Fprintf(w, "stream: %t\n", m.Stream)
	fmt.Fprintf(w, "elapsed_sec: %.3f\n", m.ElapsedSec)
	fmt.Fprintf(w, "text_chars: %d\n", m.TextChars)
	fmt.Fprintf(w, "text_tps: %.2f chars/sec\n", tps)
	fmt.Fprintf(w, "status: %s\n", m.Status)
	if m.ImageCount > 0 {
		fmt.Fprintf(w, "images: %d\n", m.ImageCount)
	}
	if m.ExceptionMessage != "" {
		fmt.Fprintf(w, "exception: %s\n", m.ExceptionMessage)
	}
}

File: client/sdk/typescript/src/metrics.ts
export type RequestMetrics = {
  provider: string;
  model: string;
  baseURL: string;
  stream: boolean;
  elapsedSec: number;
  textChars: number;
  imageCount?: number;
  status: string;
  exceptionMessage?: string;
};

export function formatRequestMetrics(metrics: RequestMetrics): string {
  const elapsed = metrics.elapsedSec > 0 ? metrics.elapsedSec : 1e-9;
  const tps = metrics.textChars / elapsed;
  const lines = [
    "",
    "=== Request Metrics ===",
    `provider: ${metrics.provider}`,
    `model: ${metrics.model}`,
    `base_url: ${metrics.baseURL}`,
    `stream: ${String(metrics.stream)}`,
    `elapsed_sec: ${metrics.elapsedSec.toFixed(3)}`,
    `text_chars: ${metrics.textChars}`,
    `text_tps: ${tps.toFixed(2)} chars/sec`,
    `status: ${metrics.status}`,
  ];

  if ((metrics.imageCount ?? 0) > 0) {
    lines.push(`images: ${metrics.imageCount}`);
  }
  if (metrics.exceptionMessage) {
    lines.push(`exception: ${metrics.exceptionMessage}`);
  }
  return `${lines.join("\n")}\n`;
}
